lib/apis/v3/bgppeer.go

```diff
@@ -44,8 +44,10 @@ type BGPPeerSpec struct {
 	// If this is not set, this represents a global peer, i.e. a peer that peers with
 	// every node in the deployment.
 	Node string `json:"node,omitempty" validate:"omitempty,name"`
-	// The IP address of the peer.
-	PeerIP string `json:"peerIP" validate:"omitempty,ip"`
+	// The IP address of the peer followed by an optional port number to peer with.
+	// If port number is given, format should be `[<IPv6>]:port` or `<IPv4>:<port>` for IPv4.
```

Comment:
yes, we would be using the `listenPort` of `BGPConfiguration` in that case.

Reply:
To check my understanding, I think you are proposing this here:
`If the PeerIP address matches another Calico node, the PeerIP port is ignored.`

If that's right, I think it's wrong, because we don't have similar logic for the `ASNumber` field: 
https://github.com/projectcalico/confd/blob/master/pkg/backends/calico/client.go#L429:L432

Note that we specifically allow for a different `ASNumber` when computing any reverse peerings:
https://github.com/projectcalico/confd/blob/master/pkg/backends/calico/client.go#L462

Also your current proposal would rule out use cases that could be useful in future, where there is another BGP daemon (than calico-node) listening on a Calico node with a different port.

Net, I suggest removing this proposal behaviour, and the "Port number used by this peer if it is targeting other calico/node will be the `listenPort` set in `BGPConfiguration`" line.

Reply:
>If the PeerIP address matches another Calico node, the PeerIP port is ignored.

I think that not the case.

My understanding of @caseydavenport comment was, say if calico/node has non-default listen_port, then the BGP peer during reverse peering will use this calico/node's ip and listen_port for peering (aka auto-detect calico/node's port number to peer with).

PeerIP port will be used only during forward peering where calico/node targets the peer, during revere peering where peers target the calico/node it will be based on the listen_port of the target node.

Reply:
Thanks @Suraiya-Hameed, that specification sounds perfect to me, so I guess I just misunderstood this sentence of the comment.  Can you think of a way to clarify it?  In particular I think "if it is targeting other calico/node" is misleading.

Reply:
I've updated one of the confd test to cover this case https://github.com/projectcalico/confd/blob/946cbe7428a6e09476fa90b00de33ce67a715187/tests/mock_data/calicoctl/explicit_peering/global/input.yaml

In the `input.yaml`, when `kube-master` is peering with `bgppeer-2`,  `bgppeer-2` has `PeerIP` without port, plus it is also a calico/node(`kube-node-2`). If `kube-node-2` has a `listenPort` set, we will use that during peering.

This will be the outcome, 
```
# For peer /global/peer_v4/10.192.0.4
protocol bgp Global_10_192_0_4 from bgp_template {
  neighbor 10.192.0.4 port 150 as 64567; 
  #`bgppeer-2` dint have any port set, but `kube-node-2` had a listenPort, so we are using that port when `kube-master` adds `bgppeer-2` as neighbour.
}
```